pkg/pillar/cmd/volumemgr/updatestatus.go

```diff
@@ -6,165 +6,279 @@ package volumemgr
+				status.DisplayName, status.ContentID,
+				status.ContentSha256)
 			return changed, false
 		}
 	}
-	if status.State == types.VERIFIED && !status.VolumeCreated {
-		status.State = types.CREATING_VOLUME
-		changed = true
-		// Asynch creation; ensure we have requested it
-		MaybeAddWorkCreate(ctx, status)
-	}
 	// If maximum download size is 0 then we are updating the
 	// downloaded size of an image in MaxSizeBytes
-	if status.DownloadOrigin.MaxDownSize == 0 {
+	if status.MaxDownloadSize == 0 {
 
 		info, err := os.Stat(status.FileLocation)
 		if err != nil {
 			errStr := fmt.Sprintf("Calculating size of container image failed: %v", err)
 			log.Error(errStr)
 		} else {
-			status.DownloadOrigin.MaxDownSize = uint64(info.Size())
+			status.MaxDownloadSize = uint64(info.Size())
 		}
 	}
-	if status.State == types.CREATING_VOLUME && !status.VolumeCreated {
-		vr := lookupVolumeWorkResult(ctx, status.Key())
-		if vr != nil {
-			log.Infof("doUpdate: VolumeWorkResult(%s) location %s, created %t",
-				status.Key(), vr.FileLocation, vr.VolumeCreated)
-			deleteVolumeWorkResult(ctx, status.Key())
-			if status.VolumeCreated != vr.VolumeCreated {
-				log.Infof("From vr set VolumeCreated to %s for %s",
-					vr.FileLocation, status.VolumeID)
-				status.VolumeCreated = vr.VolumeCreated
-				changed = true
-			}
-			if status.FileLocation != vr.FileLocation {
-				log.Infof("doUpdate: From vr set FileLocation to %s for %s",
-					vr.FileLocation, status.VolumeID)
-				status.FileLocation = vr.FileLocation
-				changed = true
-			}
-			if vr.Error != nil {
-				status.SetErrorWithSource(vr.Error.Error(),
-					types.OldVolumeStatus{}, vr.ErrorTime)
+	if status.State == types.VERIFIED {
+		return changed, true
+	}
+	return changed, false
+}
+
+// Returns changed
+// XXX remove "done" boolean return?
+func doUpdateVol(ctx *volumemgrContext, status *types.VolumeStatus) (bool, bool) {
+
+	log.Infof("doUpdateVol(%s) name %s", status.Key(), status.DisplayName)
+
+	// Anything to do?
+	if status.State == types.CREATED_VOLUME {
+		log.Infof("doUpdateVol(%s) name %s nothing to do",
+			status.Key(), status.DisplayName)
+		return false, true
+	}
+	changed := false
+	switch status.VolumeContentOriginType {
+	case zconfig.VolumeContentOriginType_VCOT_BLANK:
+		// XXX TBD
+		errStr := fmt.Sprintf("doUpdateVol(%s) name %s: Volume content origin type %v is not implemeted yet.",
+			status.Key(), status.DisplayName, status.VolumeContentOriginType)
+		status.SetErrorWithSource(errStr,
+			types.VolumeStatus{}, time.Now())
+		changed = true
+		return changed, false
+	case zconfig.VolumeContentOriginType_VCOT_DOWNLOAD:
+		ctStatus := lookupContentTreeStatus(ctx, status.ContentID.String())
+		if ctStatus == nil {
+			// Content tree not available
+			log.Infof("doUpdateVol(%s) name %s: waiting for content tree status %v",
+				status.Key(), status.DisplayName, status.ContentID)
+			return changed, false
```

Comment:
If this last forever it will just sit there without an error. I wonder if it instead makes sense to put this as an error in the volumestatus. In the normal case you'd either find the ctStatus, or it will appear very soon. Depends on the order in which volumemgr processes volumeConfig vs ContentTreeConfig from zedagent.

Reply:
Added an error state if content tree status is not available while processing volumes.